rfd/0157-aws-eks-discover.md

```diff
@@ -0,0 +1,185 @@
+We will reuse the existing EKS discovery capability, although a new mode will be added to it - discovery through
+the AWS integration. EKS discovery in that mode will rely on the AWS integration, selected in the Discovery UI,
+to find clusters and then install the Helm chart on them. The algorithm will be as follows:
+- List EKS clusters through the AWS integration.
+- Cross-check to see which EKS clusters are already registered in the system.
+- Once a new EKS cluster is found, associate our OIDC provider with it.
+- Using the associated provider, generate an access token and install the kube agent through the Helm chart.
+
+Users will be able to filter EKS clusters for discovery by using labels.
+
+A dynamic discovery configuration will be created as a result of users setting up automatic EKS discovery in the Discover UI.
+This configuration can be picked up by any discovery service. In the Teleport cloud, the Discovery service will be available by default,
+meaning users will not need to perform any additional actions.
+
+### EKS Identity Provider Association
+
+AWS EKS supports a method of authentication that involves associating an OIDC identity provider with the cluster.
+Such an associated identity provider can issue access tokens (JWT) that grant access to the EKS cluster groups and users.
+Since Teleport already supports OIDC providers, we can easily use this approach. It will require slight
+modifications to the AWS OIDC integration code to allow it to issue JWTs for an EKS cluster. To associate an OIDC provider with an
+EKS cluster, the selected AWS integration needs to issue an API call `eks:AssociateIdentityProviderConfig`
+with the following configuration (includes example parameters):
+
+```yaml
+---
+apiVersion: eksctl.io/v1alpha5
+kind: ClusterConfig
+
+metadata:
+  name: eks-cluster-name
+  region: us-east-1
+
+identityProviders:
+  - name: TeleportProvider
+    type: oidc
+    issuerUrl: https://teleport.example.com
+    clientId: kubernetes
+    usernameClaim: email
+    groupsClaim: sub
+```
+
+To access the target EKS cluster in API mode, we will generate an access token using the associated OIDC provider.
+That token will grant access to the `system:masters` group, so the Helm chart installation will be able to set up
+all the required infrastructure for the Teleport kube agent, including service accounts.
+
+## Product Usage
+
+We will add a new PostHog event to track the usage of the Discover flow for EKS.
+
+```protobuf
+// UIDiscoverEKSClusterEnrollEvent is emitted when a user is finished with
+// the step that asks user to select from a list of EKS clusters.
+message UIDiscoverEKSClusterEnrollEvent {
+  DiscoverMetadata metadata = 1;
+  DiscoverResourceMetadata resource = 2;
+  DiscoverStepStatus status = 3;
+  int64 selected_resources_count = 4;
+}
+```
+
+## Security
+
+New permissions will be required for the AWS OIDC integration to perform the necessary tasks. It will need
```

Comment:
Several prerequisites must be ensured in this context. The first requirement is that no user should have the ability to create a Token that can be used to authenticate in EKS clusters. The second requirement is that any token generated by Teleport (default OIDC provider) must not include the name of the claim designated for sending User and Kubernetes groups to the EKS clusters. 

This precaution is vital to prevent a user from circumventing Teleport by generating tokens with claims that overlap those defined in the cluster's OIDC providers and gaining access to cluster. It is advisable to explicitly define and restrict the claims to minimize potential issues. Ideally, considering the utilization of separate signers for infrastructure and user-related tasks would be beneficial.

Can you please elaborate more this security section?

Reply:
Yep, as discussed offline, I added more information about these tokens to the security session, their separate CA usage, audience field and specific user/group claim names 👍